service-api/module/Application/src/Controller/YotiController.php

```diff
@@ -107,10 +128,58 @@ public function getSessionStatusAction(): JsonModel
         return new JsonModel($data);
     }
 
-    public function getPDFLetterAction(string $session): JsonModel
+    /**
+     * @throws YotiException
+     */
+    public function preparePDFLetterAction(): JsonModel
     {
+        $uuid = $this->params()->fromRoute('uuid');
+        if (! $uuid) {
+            $this->getResponse()->setStatusCode(Response::STATUS_CODE_400);
+            return new JsonModel(['error' => 'Missing uuid']);
+        }
+
+        $caseData = $this->dataQuery->getCaseByUUID($uuid);
+
+        if (! $caseData || $caseData->sessionId === null) {
+            $this->getResponse()->setStatusCode(Response::STATUS_CODE_400);
+            return new JsonModel(new Problem('SessionId does not exist to prepare PDF'));
+        }
+
         $data = [];
-        $data['response'] = $this->yotiService->retrieveLetterPDF($session);
+        try {
+            $data['response'] = $this->yotiService->preparePDFLetter($caseData);
+        } catch (YotiException $e) {
+            $this->getResponse()->setStatusCode(Response::STATUS_CODE_500);
+            return new JsonModel(new Problem(
+                'Problem preparing PDF Yoti API',
+                extra: ['errors' => $e->getMessage()],
+            ));
+        }
+
+        $this->getResponse()->setStatusCode(Response::STATUS_CODE_200);
         return new JsonModel($data);
     }
+
+    public function retrievePDFAction(): JsonModel
+    {
+        $uuid = $this->params()->fromRoute('uuid');
+        $caseData = $this->dataQuery->getCaseByUUID($uuid);
+        try {
+            $data = $this->yotiService->retrieveLetterPDF($caseData);
+        } catch (YotiException $e) {
+            $this->getResponse()->setStatusCode(Response::STATUS_CODE_500);
+            return new JsonModel(new Problem(
+                'Problem retrieving PDF Yoti API',
+                extra: ['errors' => $e->getMessage()],
+            ));
+        }
+        // Write pdf to file
+        $fileName = 'instructions-' . $uuid . '.pdf';
+        $content = file_put_contents($fileName, $data['pdfData']);
```

Comment:
Any particular reason to save to the file system? This can leave PII lying about on containers, so is not really advisable and it looks like you are just sending pdfData onward to the response anyway.  If you need to send the file-name onward, you can use a [Content Disposition](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition) header.

Reply:
This was just so I could see the pdf locally, will remove, as just passing it onwards to caller as response